modules/auxiliary/scanner/smb/smb_version.rb

```diff
@@ -39,167 +50,250 @@ def smb_direct
+    end
+
+    { text: desc, fingerprint_match: nd_fingerprint_match, smb_fingerprint: nd_smb_fingerprint }
+  end
+
+  #
+  # Fingerprint a single host
+  #
+  def run_host(ip)
+    smb_ports = [445, 139]
+    lines = [] # defer status output to the very end to group lines together by host
+    smb_ports.each do |pnum|
+      @smb_port = pnum
+      self.simple = nil
 
-        if res['lang'].to_s.length > 0 and res['lang'] != 'Unknown'
-          desc << " (language:#{res['lang']})"
-          conf[:os_lang] = res['lang']
-          match_conf['os.language'] = conf[:os_lang]
+      begin
+        res = smb_fingerprint
+
+        info = smb_proto_info
+        desc = "SMB Detected (versions:#{info[:versions].join(', ')}) (preferred dialect:#{info[:preferred_dialect]})"
+        info[:capabilities].each do |name, values|
+          desc << " (#{name} capabilities:#{values.join(', ')})"
         end
 
-        if simple.client.default_name
-          desc << " (name:#{simple.client.default_name})"
-          conf[:SMBName] = simple.client.default_name
-          match_conf['host.name'] = conf[:SMBName]
+        if simple.client.peer_require_signing
+          desc << ' (signatures:required)'
+        else
+          desc << ' (signatures:optional)'
+          report_vuln({
+            host: ip,
+            port: rport,
+            proto: 'tcp',
+            name: 'SMB Signing Is Not Required',
+            refs: [
+              SiteReference.new('URL', 'https://support.microsoft.com/en-us/help/161372/how-to-enable-smb-signing-in-windows-nt'),
+              SiteReference.new('URL', 'https://support.microsoft.com/en-us/help/887429/overview-of-server-message-block-signing'),
+            ]
+          })
         end
+        desc << " (uptime:#{info[:uptime]})" if info[:uptime]
+        desc << " (guid:#{Rex::Text.to_guid(info[:server_guid])})" if info[:server_guid]
+        lines << { type: :status, message: desc }
 
-        if simple.client.default_domain
-          if simple.client.default_domain.encoding.name == "UTF-8"
-            desc << " (domain:#{simple.client.default_domain})"
-          else
-            # Workgroup names are in ANSI, but may contain invalid characters
-            # Go through each char and convert/check
-            temp_workgroup = simple.client.default_domain.dup
-            desc << " (workgroup:"
-            temp_workgroup.each_char do |i|
-              begin
-                desc << i.encode("UTF-8")
-              rescue Encoding::UndefinedConversionError => e
-                desc << '?'
-                print_error("Found incompatible (non-ANSI) character in Workgroup name. Replaced with '?'")
+        #
+        # Create the note hash for smb.fingerprint
+        #
+        nd_smb_fingerprint = {
+          native_os: res['native_os'],
+          native_lm: res['native_lm']
+        }
+
+        if res['os'] && res['os'] != 'Unknown'
```

Comment:
We can still get some information with SMBv2/3, even if SMBv1 is disabled. By trying to authenticate, the `NTLM Type 2` message received when trying to authenticate contains some information that might be useful. One of them is `default_domain`, that is used below, and sometimes not provided by the Rex client. RubySMB parses it [here](https://github.com/rapid7/ruby_smb/blob/master/lib/ruby_smb/client/authentication.rb#L336).

I did some tests and here are the results:
```
    206:         require 'pry';binding.pry
 => 207:         if res['os'] && res['os'] != 'Unknown'
    208:           description = smb_os_description(res, nd_smb_fingerprint)
    209:           desc = description[:text]
    210:           nd_fingerprint_match = description[:fingerprint_match]
    211:           nd_smb_fingerprint = description[:smb_fingerprint]
    212:

[1] pry(#<Msf::Modules::Auxiliary__Scanner__Smb__Smb_version::MetasploitModule>)> res
=> {"native_os"=>nil, "native_lm"=>nil, "os"=>"Unknown", "sp"=>"", "lang"=>"Unknown"}
[2] pry(#<Msf::Modules::Auxiliary__Scanner__Smb__Smb_version::MetasploitModule>)> connect(versions:[1,2,3])
=> #<Rex::Proto::SMB::SimpleClient:0x00007fb732ea1140
 @client=
   ...[SNIP]...
[3] pry(#<Msf::Modules::Auxiliary__Scanner__Smb__Smb_version::MetasploitModule>)> simple.client.negotiate
=> "SMB3"
[4] pry(#<Msf::Modules::Auxiliary__Scanner__Smb__Smb_version::MetasploitModule>)> simple.client.authenticate
=> #<WindowsError::ErrorCode:0x00007fb7328ebbb0 @description="The attempted logon is invalid. This is either due to a bad username or authentication information.", @name="STATUS_LOGON_FAILURE", @value=3221225581>
[5] pry(#<Msf::Modules::Auxiliary__Scanner__Smb__Smb_version::MetasploitModule>)> simple.client.default_domain
=> "EC2AMAZ-7LOHKG5"
```

Reply:
Also, I noticed that, when the Rex client gets the `default_domain` on older Windows (Windows 2008 R2), it looks like to be the workstation instead:
```
    206:         require 'pry';binding.pry
 => 207:         if res['os'] && res['os'] != 'Unknown'
    208:           description = smb_os_description(res, nd_smb_fingerprint)
    209:           desc = description[:text]
    210:           nd_fingerprint_match = description[:fingerprint_match]
    211:           nd_smb_fingerprint = description[:smb_fingerprint]
    212:

[1] pry(#<Msf::Modules::Auxiliary__Scanner__Smb__Smb_version::MetasploitModule>)> simple.client.default_domain
=> "WORKGROUP"
[2] pry(#<Msf::Modules::Auxiliary__Scanner__Smb__Smb_version::MetasploitModule>)> connect(versions:[1,2,3])
=> #<Rex::Proto::SMB::SimpleClient:0x00007fb7290c6ba0
 @client=
  #<RubySMB::Client:0x00007fb7290c6808
   ...[SNIP]...
[3] pry(#<Msf::Modules::Auxiliary__Scanner__Smb__Smb_version::MetasploitModule>)> simple.client.negotiate
=> "SMB2"
[4] pry(#<Msf::Modules::Auxiliary__Scanner__Smb__Smb_version::MetasploitModule>)> simple.client.authenticate
=> #<WindowsError::ErrorCode:0x00007fb7328ebbb0 @description="The attempted logon is invalid. This is either due to a bad username or authentication information.", @name="STATUS_LOGON_FAILURE", @value=3221225581>
[5] pry(#<Msf::Modules::Auxiliary__Scanner__Smb__Smb_version::MetasploitModule>)> simple.client.default_domain
=> "WIN-RHDA4UK6PF6"
```

Reply:
Oh, I just saw that Rex client has some interesting logic for that: https://github.com/rapid7/metasploit-framework/blob/master/lib/rex/proto/smb/client.rb#L926-L938
```ruby
    blob_data = NTLM_UTILS.parse_ntlm_type_2_blob(blob)
    #netbios name
    self.default_name =  blob_data[:default_name] || ''
    #dns name
    self.dns_host_name =  blob_data[:dns_host_name] || ''
    #dns domain
    if blob_data[:default_name] != blob_data[:default_domain]
      # We're in a domain; get the domain name now
      self.default_domain =  blob_data[:default_domain] || ''
    else
      # We're in a workgroup; workgroup names come later in the handshake
      self.default_domain = nil
    end
```
We might need to update RubySMB to match this.

Reply:
So, maybe `default_name ` should be used for workstation and `default_domain`, if not nil, for domain.

Reply:
You're right, that information about the domain should not get lost. I added the ability to log the `authentication domain` in commit 42aa479ef2f370f11fb65ed21dd6b8faf6615db5. By putting it at the SMB protocol level instead of treating it as host fingerprint information, it shows up even when the fingerprinting fails as you can see in the output below when I scanned my Linux workstation 192.168.159.128.

New output:
```
[*] 192.168.159.10:445    - SMB Detected (versions:1, 2, 3) (preferred dialect:SMB 3.1.1) (compression capabilities:LZNT1) (encryption capabilities:AES-128-CCM) (signatures:required) (guid:{faf5534c-d125-4081-aa2a-cf3256415908}) (authentication domain:MSFLAB)
[*] 192.168.159.10:445    -   Host could not be identified: Windows Server 2019 Standard 17763 (Windows Server 2019 Standard 6.3)
[*] 192.168.159.0-130:    - Scanned  16 of 131 hosts (12% complete)
[*] 192.168.159.0-130:    - Scanned  27 of 131 hosts (20% complete)
[*] 192.168.159.31:445    - SMB Detected (versions:1, 2) (preferred dialect:SMB 2.1) (signatures:optional) (uptime:1h 35m 49s) (guid:{057ab4e7-06bf-455a-8c9a-7136889c6f5a}) (authentication domain:WIN-9NSI4A6AIHJ)
[+] 192.168.159.31:445    -   Host is running Windows 7 Professional SP1 (build:7601) (name:WIN-9NSI4A6AIHJ) (workgroup:WORKGROUP)
[*] 192.168.159.38:445    - SMB Detected (versions:2, 3) (preferred dialect:SMB 3.0.2) (signatures:optional) (uptime:7m 52s) (guid:{cd5d41db-0bb8-4954-9421-0cdd14b7c6f7}) (authentication domain:WIN-46IL3RC2FHI)
[*] 192.168.159.0-130:    - Scanned  42 of 131 hosts (32% complete)
[*] 192.168.159.0-130:    - Scanned  57 of 131 hosts (43% complete)
[*] 192.168.159.48:445    - SMB Detected (versions:1) (preferred dialect:NT LM 0.12) (signatures:optional)
[+] 192.168.159.48:445    -   Host is running Windows XP SP2 (language:English) (name:SMCINTYR-81CC7C) (workgroup:WORKGROUP)
[*] 192.168.159.0-130:    - Scanned  72 of 131 hosts (54% complete)
[*] 192.168.159.0-130:    - Scanned  87 of 131 hosts (66% complete)
[*] 192.168.159.0-130:    - Scanned  92 of 131 hosts (70% complete)
[*] 192.168.159.0-130:    - Scanned 107 of 131 hosts (81% complete)
[*] 192.168.159.128:445   - SMB Detected (versions:2, 3) (preferred dialect:SMB 3.1.1) (compression capabilities:LZ77) (encryption capabilities:AES-128-GCM) (signatures:optional) (guid:{61636f6c-686c-736f-7400-000000000000}) (authentication domain:LOCALHOST)
[*] 192.168.159.129:445   - SMB Detected (versions:1, 2, 3) (preferred dialect:SMB 3.1.1) (compression capabilities:LZNT1) (encryption capabilities:AES-128-CCM) (signatures:optional) (guid:{423637e5-d8a9-407a-ab7f-4d1879dc1772}) (authentication domain:DESKTOP-R9TM84E)
[+] 192.168.159.129:445   -   Host is running Windows 10 Enterprise (build:17763) (name:DESKTOP-R9TM84E) (workgroup:WORKGROUP)
[*] 192.168.159.0-130:    - Scanned 123 of 131 hosts (93% complete)
[*] 192.168.159.0-130:    - Scanned 131 of 131 hosts (100% complete)
[*] Auxiliary module execution completed
```